### PR TITLE
Deprecate chiselName and NoChiselNamePrefix trait

### PR DESCRIPTION
**Pull request by jared-barocsi**
This PR deprecates `@chiselName` and `NoChiselNamePrefix` in favor of the newly added `AffectsChiselPrefix` trait, and deletes all non-testing uses of them.

Needs to be backported so that a followup PR that removes both of these can be made

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- code cleanup/refactoring

#### API Impact
- deprecates chisellName and NoChiselNamePrefix

#### Backend Code Generation Impact
- no impact

#### Desired Merge Strategy
- squash and merge

#### Release Notes
- Deprecate chiselName and associated NoChiselNamePrefix trait

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/main/scala/chisel3/util/Counter.scala [hunk 1]
```diff
@@ -113,7 +112,6 @@ object Counter {
     * @return tuple of the counter value and whether the counter will wrap (the value is at
     * maximum and the condition is true).
     */
-  @chiselName
   def apply(cond: Bool, n: Int): (UInt, Bool) = {
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Counter.scala:115]: does returning Tuples of hardware work with the plugin?

File: src/test/scala/chiselTests/StrongEnum.scala [hunk 2]
```diff
@@ -575,11 +575,10 @@ class StrongEnumAnnotator extends Module {
   val indexed2 = vec_of_bundles(cycle)
 }
 
-@chiselName
 class StrongEnumAnnotatorWithChiselName extends Module {
   import EnumExample._
 
-  object LocalEnum extends ChiselEnum {
+  object LocalEnum extends ChiselEnum with AffectsChiselPrefix {
```
Comment (mwachs5) [inline src/test/scala/chiselTests/StrongEnum.scala:581]: should `ChiselEnum` `AffectsChiselPrefix`?

Comment (mwachs5) [reply src/test/scala/chiselTests/StrongEnum.scala:581]: ah i see what you mean, agree that maybe making ChiselEnums always be named is OK to do later

File: src/main/scala/chisel3/util/Counter.scala (hunk 1 above)
Comment (jared-barocsi) [reply src/main/scala/chisel3/util/Counter.scala:115]: I don't think so, but I'm not sure how to return a Tuple that extends `AffectsChiselPrefix`

Comment (jared-barocsi) [reply src/main/scala/chisel3/util/Counter.scala:115]: @jackkoenig Any ideas?

Comment (jared-barocsi) [reply src/main/scala/chisel3/util/Counter.scala:115]: @mwachs5 I created a module that calls this function, with and without `chiselName` and there does not seem to be any difference in the naming of the resulting firrtl. I don't believe that `chiselName` is doing anything here.

**Review by jackkoenig — APPROVED** [review]

**Comment by pxs7** [comment]
when compile hardfloat， 
 it note，    AffectsChiselPrefix is not a member of package chisel3.experimental， i also  search for  AffectsChiselPrefix  in chisel3 code， not such member？

**Pull request merged**